libpod/runtime_img.go

```diff
@@ -276,46 +278,112 @@ func DownloadFromFile(reader *os.File) (string, error) {
+}
+
+// LoadImageFromSingleImageDir loads image from the archive of single image that inputFile points to.
+func (r *Runtime) LoadImageFromSingleImageDir(ctx context.Context, writer io.Writer, inputFile, signaturePolicy, format string) (string, error) {
+	var (
+		src types.ImageReference
+		err error
+	)
+	switch format {
+	case image.DirTransport:
+		if src, err = directory.NewReference(inputFile); err != nil {
+			return "", err
+		}
+	case image.OCIDirTransport:
+		if src, err = layout.NewReference(inputFile, ""); err != nil {
+			return "", err
+		}
+	}
+	if src != nil {
+		newImages, err := r.ImageRuntime().LoadFromArchiveReference(ctx, src, signaturePolicy, writer)
+		if err != nil {
+			return "", err
+		}
 		return getImageNames(newImages), nil
 	}
-	return "", err
+
+	return "", errors.Wrapf(err, "error pulling image")
 }
 
-// LoadImageFromSingleImageArchive load image from the archive of single image that inputFile points to.
-func (r *Runtime) LoadImageFromSingleImageArchive(ctx context.Context, writer io.Writer, inputFile, signaturePolicy string) (string, error) {
-	var err error
-	for _, referenceFn := range []func() (types.ImageReference, error){
-		func() (types.ImageReference, error) {
-			return dockerarchive.ParseReference(inputFile)
-		},
-		func() (types.ImageReference, error) {
-			return ociarchive.NewReference(inputFile, "")
-		},
-		func() (types.ImageReference, error) {
-			return directory.NewReference(inputFile)
-		},
-		func() (types.ImageReference, error) {
-			return layout.NewReference(inputFile, "")
-		},
-	} {
-		src, err := referenceFn()
-		if err == nil && src != nil {
-			if newImages, err := r.ImageRuntime().LoadFromArchiveReference(ctx, src, signaturePolicy, writer); err == nil {
-				return getImageNames(newImages), nil
+// imageFormat detects the image foramt of the input archive or directory
+// returns "", if can not decide the format
+func imageFormat(inputFile string) (string, error) {
+	path, err := explicitfilepath.ResolvePathToFullyExplicit(inputFile)
+	if err != nil {
+		return "", errors.Wrapf(err, "error detecting input file format")
+	}
+	if !archive.IsArchivePath(inputFile) {
+		files, err := ioutil.ReadDir(path)
+		if err != nil {
+			return "", errors.Wrapf(err, "error detecting input file format")
+		}
+		for _, f := range files {
+			if f.Name() == "manifest.json" {
+				return image.DirTransport, nil
+			}
+			if f.Name() == "index.json" {
+				return image.OCIDirTransport, nil
 			}
 		}
+		return "", errors.Wrapf(err, "error detecting input file format")
 	}
-	return "", errors.Wrapf(err, "error pulling image")
+	f, err := os.Open(path)
+	if err != nil {
+		return "", errors.Wrapf(err, "error detecting input file format")
+	}
+	reader := tar.NewReader(f)
+	for {
+		header, err := reader.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return "", errors.Wrapf(err, "error detecting archive format")
```

Comment:
Is this the path taken on a compressed archive? That should probably return `"", nil` (and the caller must handle that) to indicate that it didn’t make a guess.